dashproducts.php

```diff
@@ -200,7 +200,7 @@ public function getTableBestSellers($date_from, $date_to)
 					SELECT
 						product_id,
 						product_name,
-						SUM(product_quantity-product_quantity_refunded-product_quantity_return-product_quantity_reinjected) as total,
+						SUM(CAST(product_quantity as signed)-CAST(product_quantity_refunded as signed)-CAST(product_quantity_return as signed)-CAST(product_quantity_reinjected as signed)) as total,
```

Comment:
As far as I know, is not CAST function only admit the value and the type? You cannot put a formula inside, isn't it?

Also, with the CAST as signed, per value, you don't have the problem about "negative" ones, because the cast is as signed (this is, with the proper negative or positive sign).

Reply:
Another test, that could be valid the syntax for the CAST is:
`SUM(CAST(product_quantity-product_quantity_refunded-product_quantity_return-product_quantity_reinjected as signed))  as total`

But then, it also rises:
[Wed Jan 26 09:37:38.229813 2022] [proxy_fcgi:error] [pid 1174986:tid 139922702849792] [client 162.158.50.33:0] AH01071: Got error 'PHP message: BIGINT UNSIGNED value is out of range in '`prestashop_db`.`od`.`product_quantity` - `prestashop_db`.`od`.`product_quantity_refunded` - `prestashop_db`.`od`.`product_quantity_return` - `prestashop_db`.`od`.`product_quantity_reinjected`'PHP message: PHP Warning:  Invalid argument supplied for foreach() in /var/www/vhosts/domain.com/httpdocs/modules/dashproducts/dashproducts.php on line 219', referer: https://www.domain.com/backoffice/index.php?controller=AdminDashboard&token=f22f4c6102405b64b02dd1c5d177932e

So the only solution is casting each value.

Reply:
Hi @PierreRambaud , here is the explanation, and is exactly the problem: https://es.stackoverflow.com/questions/127854/mysql-error-1690-bigint-unsigned-value-is-out-of-range-in

So, on the original code, the substract inside the "SUM as total" without CAST properly each value when you are substracting all the values expect to be a positive number, but this is clearly false because stocks can be negative then the result can be negative. An unsigned integer can be only 0 to 18446744073709551615. So, the result of that is "BIGINT UNSIGNED value out of range" exception, is cause produce an overflow exception due the substraction could be perfectly a SIGNED ( < 0) value.

For that, like the example on stackoverflow, you must enclose each value on the substraction casting as signed, at least the ones that for sure can be a negative value. So with that, casting each value as signed, we can have values for each ones from -9223372036854775808 to 9223372036854775807.